documentation/library_guide/kernel_templates/esimd/radix_sort.rst

```diff
@@ -0,0 +1,185 @@
+
+
+**Limitations (may be relaxed in the future)**
+
+- Algorithms can process only C++ integral and floating-point types with the width up to 64-bits (except for ``bool``).
+- Number of elements to sort must not exceed `2^30`.
+- ``radix_bits`` can only be `8`.
+- ``param.data_per_workitem`` has discreteness of `32`.
+- ``param.workgroup_size`` can be either `32` or `64`.
+- Local memory is always used to rank keys, reorder keys or key-value pairs which limits possible values of ``param.data_per_workitem`` and ``param.workgroup_size``.
+- ``radix_sort_by_key`` does not have single-work-group implementation yet.
+
+
+**Possible API extensions (may be implemented in the future)**
+
+- Allow passing externally allocated memory.
+- Allow passing dependent events.
+- Allow passing a range of bits to sort.
+- Allow out-of-place sorting, e.g. with a double-buffer or an output sequence(s)
+- Allow configuration of kernels other than the most time-consuming kernel (e.g. of a kernel computing histograms).
+- Allow range transformations (e.g. range pipes or transform iterators).
+
+
+**System requirements (coverage my be extended in the future)**
+
+- Hardware: Intel® Data Center GPU Max Series.
+- Compiler: Intel® oneAPI DPC++/C++ 2023.2 and newer.
+- OS: RHEL 9.2, SLES 15 SP5, Ubuntu 22.04. Other distributions and their versions listed in `<https://dgpu-docs.intel.com/driver/installation.html>` should be supported accordingly.
+
+
+**Known Issues**
+
+- Use of -g, -O0, -O1 compiler options may lead to compilation issues.
+- Combinations of ``param.data_per_workitem`` and ``param.work_group_size`` with large values may lead to device-code compilation errors due to allocation of local memory amounts beyond the device capabilities. Refer to "Local memory usage" paragraph for the details regarding allocation.
+- Some combinations of types and ``kt::kernel_param`` values lead to wrong results starting with `20231219 <https://dgpu-docs.intel.com/releases/stable_775_20_20231219.html>`_ rolling release of the GPU driver:
+
+  - ``radix_sort`` with ``sizeof(key_type)=8``, ``param.workgroup_size = 32``, and ``param.data_per_workitem>=288``
+  - ``radix_sort_by_key`` with ``4 <= sizeof(key_type) + sizeof(value_type) <= 8``, ``param.workgroup_size = 32`` and ``param.data_per_workitem >= 288``
+  - ``radix_sort_by_key`` with ``9 <= sizeof(key_type) + sizeof(value_type) <= 10``, ``param.workgroup_size = 32`` and ``param.data_per_workitem >= 224``
+  - ``radix_sort_by_key`` with ``sizeof(key_type) + sizeof(value_type) > 10``, ``param.workgroup_size = 32`` and ``param.data_per_workitem = 64``
+  - ``radix_sort_by_key`` with ``sizeof(key_type) + sizeof(value_type) = 12``, ``param.workgroup_size = 64`` and ``param.data_per_workitem = 96``
+  - ``radix_sort_by_key`` with ``sizeof(key_type) + sizeof(value_type) = 16``, ``param.workgroup_size = 64`` and ``param.data_per_workitem = 64``
```

Comment:
Given the results of the testing with the latest rolling release, I am considering setting "32" as a non-supported value for workgroup_size, because there are too many failures is no a simple pattern to describe them, despite the fact that the overall pass rate is ~94%.

Below are the tables with the test results. DPWI is data_per_workitem, WGS is workgroup_size, 1B is 1 byte data, 2B is 2 byte data type and so on. Different types with the same number of bytes (e.g. double and uint64_t), different key-value combinations (e.g. 2B-key and 4B-value, 4B-key and 2B-value) are grouped together: failure in one test is shown as failure of all the combinations.

``radix_sort``, ``WGS=32``:

DPWI | WGS | 1B | 2B | 4B | 8B
-- | -- | -- | -- | -- | --
32 | 32 | pass | pass | pass | pass
64 | 32 | pass | pass | pass | pass
96 | 32 | pass | pass | pass | pass
128 | 32 | pass | pass | pass | pass
160 | 32 | pass | pass | pass | pass
192 | 32 | pass | pass | pass | pass
224 | 32 | pass | pass | pass | pass
256 | 32 | pass | pass | pass | pass
288 | 32 | pass | pass | pass | fail
320 | 32 | pass | pass | pass | fail
352 | 32 | pass | pass | pass | fail
384 | 32 | pass | pass | pass | fail
416 | 32 | pass | pass | pass | fail
448 | 32 | pass | pass | pass | fail
480 | 32 | pass | pass | pass | fail
512 | 32 | pass | pass | pass | skip



``radix_sort``, ``WGS=64``:

DPWI | WGS | 1B | 2B | 4B | 8B
-- | -- | -- | -- | -- | --
32 | 64 | pass | pass | pass | pass
64 | 64 | pass | pass | pass | pass
96 | 64 | pass | pass | pass | pass
128 | 64 | pass | pass | pass | pass
160 | 64 | pass | pass | pass | pass
192 | 64 | pass | pass | pass | pass
224 | 64 | pass | pass | pass | pass
256 | 64 | pass | pass | pass | skip
288 | 64 | pass | pass | pass | skip
320 | 64 | pass | pass | pass | skip
352 | 64 | pass | pass | pass | skip
384 | 64 | pass | pass | pass | skip
416 | 64 | pass | pass | pass | skip
448 | 64 | pass | pass | pass | skip
480 | 64 | pass | pass | pass | skip
512 | 64 | pass | pass | skip | skip


``radix_sort_by_key``, ``WGS=32``:

DPWI | WGS | 2B (1B+1B) | 3B (2B+1B) | 4B (2B+2B) | 5B (32B+8B) | 6B (4B+2B) | 8B (4B+4B) | 9B (8B+1B) | 10B (8B+2B) | 12B (8B+4B) | 16B (8B+8B)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
32 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | pass | pass
64 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | pass | fail
96 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | fail | fail
128 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | pass | pass
160 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | pass | pass
192 | 32 | pass | pass | pass | pass | pass | pass | pass | pass | pass | fail
224 | 32 | pass | pass | pass | pass | pass | pass | fail | fail | fail | skip
256 | 32 | pass | pass | pass | pass | pass | pass | fail | fail | fail | skip
288 | 32 | pass | pass | pass | pass | pass | pass | fail | fail | fail | skip
320 | 32 | pass | pass | pass | pass | fail | pass | fail | fail | pass | skip
352 | 32 | pass | pass | pass | pass | fail | pass | fail | fail | skip | skip
384 | 32 | pass | pass | fail | fail | fail | fail | fail | fail | skip | skip
416 | 32 | pass | pass | fail | pass | pass | fail | fail | skip | skip | skip
448 | 32 | pass | pass | fail | fail | fail | fail | skip | skip | skip | skip
480 | 32 | pass | pass | fail | pass | fail | fail | skip | skip | skip | skip
512 | 32 | pass | pass | pass | fail | pass | skip | skip | skip | skip | skip

``radix_sort_by_key``, ``WGS=64``:

DPWI | WGS | 2B (1B+1B) | 3B (2B+1B) | 4B (2B+2B) | 5B (32B+8B) | 6B (4B+2B) | 8B (4B+4B) | 9B (8B+1B) | 10B (8B+2B) | 12B (8B+4B) | 16B (8B+8B)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
32 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | pass | pass
64 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | pass | fail
96 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | fail | pass
128 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | pass | skip
160 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | pass | skip
192 | 64 | pass | pass | pass | pass | pass | pass | pass | pass | skip | skip
224 | 64 | pass | pass | pass | pass | pass | pass | pass | skip | skip | skip
256 | 64 | pass | pass | pass | pass | pass | skip | skip | skip | skip | skip
288 | 64 | pass | pass | pass | pass | pass | skip | skip | skip | skip | skip
320 | 64 | pass | pass | pass | pass | pass | skip | skip | skip | skip | skip
352 | 64 | pass | pass | pass | pass | skip | skip | skip | skip | skip | skip
384 | 64 | pass | pass | pass | skip | skip | skip | skip | skip | skip | skip
416 | 64 | pass | pass | pass | skip | skip | skip | skip | skip | skip | skip
448 | 64 | pass | pass | pass | skip | skip | skip | skip | skip | skip | skip
480 | 64 | pass | pass | skip | skip | skip | skip | skip | skip | skip | skip
512 | 64 | pass | pass | skip | skip | skip | skip | skip | skip | skip | skip